src/banded/bidiagonalconjugation.jl

```diff
@@ -0,0 +1,141 @@
+@inline function _to_uplo(char::Symbol)
+    if char == :U
+        'U'
+    elseif char == :L
+        'L'
+    else
+        _throw_uplo()
+    end
+end
+@inline function _to_uplo(char::Char)
+    if char ∈ ('L', 'U')
+        char
+    else
+        _throw_uplo()
+    end
+end
+@noinline _throw_uplo() = throw(ArgumentError("uplo argument must be either :U (upper) or :L (lower)"))
+
+mutable struct BidiagonalConjugationData{T}
+    const U::AbstractMatrix{T} # Typing these concretely prevents the use of Bidiagonal, unless we want LazyBandedMatrices.Bidiagonal
+    const C::AbstractMatrix{T} # Function barriers help to minimise the penalty from this when resizing anyway.
+    const dv::Vector{T}
+    const ev::Vector{T}
+    const uplo::Char
+    datasize::Int # Number of columns
+end
+function BidiagonalConjugationData(U, X, V, uplo::Char)
+    C = X * V
+    T = promote_type(typeof(inv(U[1, 1])), eltype(U), eltype(C)) # include inv so that we can't get Ints
+    dv, ev = T[], T[]
+    return BidiagonalConjugationData(U, C, dv, ev, uplo, 0)
+end
+
+function copy(data::BidiagonalConjugationData)
+    U, C, dv, ev, uplo, datasize = data.U, data.C, data.dv, data.ev, data.uplo, data.datasize
+    return BidiagonalConjugationData(copy(U), copy(C), copy(dv), copy(ev), uplo, datasize)
+end
+
+function _compute_column_up!(data::BidiagonalConjugationData, U, C, i)
+    dv, ev = data.dv, data.ev
+    if i == 1
+        dv[i] = C[1, 1] / U[1, 1]
+    else
+        uᵢ₋₁ᵢ₋₁, uᵢᵢ₋₁, uᵢ₋₁ᵢ, uᵢᵢ = U[i-1, i-1], U[i, i-1], U[i-1, i], U[i, i]
+        cᵢ₋₁ᵢ, cᵢᵢ = C[i-1, i], C[i, i]
+        Uᵢ⁻¹ = inv(uᵢ₋₁ᵢ₋₁ * uᵢᵢ - uᵢ₋₁ᵢ * uᵢᵢ₋₁)
+        dv[i] = Uᵢ⁻¹ * (uᵢ₋₁ᵢ₋₁ * cᵢᵢ - uᵢᵢ₋₁ * cᵢ₋₁ᵢ)
+        ev[i-1] = Uᵢ⁻¹ * (uᵢᵢ * cᵢ₋₁ᵢ - uᵢ₋₁ᵢ * cᵢᵢ)
+    end
+    return data
+end
+
+function _compute_column_lo!(data::BidiagonalConjugationData, U, C, i)
+    dv, ev = data.dv, data.ev
+    uᵢᵢ, uᵢ₊₁ᵢ, uᵢᵢ₊₁, uᵢ₊₁ᵢ₊₁ = U[i, i], U[i+1, i], U[i, i+1], U[i+1, i+1]
+    cᵢᵢ, cᵢ₊₁ᵢ = C[i, i], C[i+1, i]
+    Uᵢ⁻¹ = inv(uᵢᵢ * uᵢ₊₁ᵢ₊₁ - uᵢᵢ₊₁ * uᵢ₊₁ᵢ)
+    dv[i] = Uᵢ⁻¹ * (uᵢ₊₁ᵢ₊₁ * cᵢᵢ - uᵢᵢ₊₁ * cᵢ₊₁ᵢ)
+    ev[i] = Uᵢ⁻¹ * (uᵢᵢ * cᵢ₊₁ᵢ - uᵢ₊₁ᵢ * cᵢᵢ)
+    return data
+end
+
+function _compute_columns!(data::BidiagonalConjugationData, i)
+    ds = data.datasize
+    up = data.uplo == 'U'
+    U, C = data.U, data.C # Treat _compute_column_(up/lo) as function barriers and take these out early
+    return __compute_columns!(data, U, C, i)
+end
+function __compute_columns!(data::BidiagonalConjugationData, U, C, i)
+    ds = data.datasize 
+    up = data.uplo == 'U'
+    for j in (ds+1):i
+        up ? _compute_column_up!(data, U, C, j) : _compute_column_lo!(data, U, C, j)
+    end
+    data.datasize = i
+    return data
+end
+
+function resizedata!(data::BidiagonalConjugationData, n)
+    ds = data.datasize
+    n ≤ ds && return data
+    dv, ev = data.dv, data.ev
+    resize!(dv, 2n + 1)
```

Comment:
I believe doing this may cause it to be an `O(n^2)` resize: if it has to copy memory everytime we call `resize!` then something like
```julia
for k=1:n
   resizedata!(data, k)
end
```
will be $O(n^2)$. 

In `CachedArray` this is avoided by doubling the size whenever we call `resize!`, see

https://github.com/JuliaArrays/LazyArrays.jl/blob/f2ed39925bad206351cd4fd7c0978310aacd52cc/src/cache.jl#L238

This is why we need `datasize` instead of just relying on the size of the array.

Reply:
Maybe I'm not understanding `_vec_resizedata!` correctly: Aren't I already doubling the size of the array here? That's what I was trying to do with the `2n` instead of just `n`

Reply:
The main difference is `_vec_resizedata!` only calls `resize!` when `n` is bigger than the length. Your version is always calling `resize!` so will be $O(n^2)$

Reply:
Doesn't `n ≤ ds && return data` make sure that no resizing actually happens when `n` exceeds the length?

Reply:
But `ds` is not the same as `length(dv)`

Reply:
I'm a bit confused on how I can change this to address what you mean. Maybe it's obvious but I can't see it sorry. From what I can see, it's resizing as I would expect. 

Here's an example. I put a `@show n` in `resizedata!` after my check `n <= ds` to see exactly when a resize is actually being performed. When resizing `dv`:

```julia
V, A, X = InfRandTridiagonal(), InfRandBidiagonal('U'), brand(∞, 0, 2);
U = X * V * ApplyArray(inv, A);
B = BidiagonalConjugation(U, X, V, 'U');
dv, ev = B.dv, B.ev; 
for n in 1:24
    println("Loop iter dv: $n")
    LazyArrays.resizedata!(dv, n)
end
```
```
Loop iter dv: 1
n = 1
Loop iter dv: 2
Loop iter dv: 3
Loop iter dv: 4
n = 4
Loop iter dv: 5
Loop iter dv: 6
Loop iter dv: 7
Loop iter dv: 8
Loop iter dv: 9
Loop iter dv: 10
n = 10
Loop iter dv: 11
Loop iter dv: 12
Loop iter dv: 13
Loop iter dv: 14
Loop iter dv: 15
Loop iter dv: 16
Loop iter dv: 17
Loop iter dv: 18
Loop iter dv: 19
Loop iter dv: 20
Loop iter dv: 21
Loop iter dv: 22
n = 22
Loop iter dv: 23
Loop iter dv: 24
```

When resizing `ev`:
```julia
B = BidiagonalConjugation(U, X, V, 'U');
dv, ev = B.dv, B.ev; 
for n in 1:24
    println("Loop iter: ev $n")
    LazyArrays.resizedata!(ev, n)
end
```
```
Loop iter: ev 1
n = 1
Loop iter: ev 2
Loop iter: ev 3
Loop iter: ev 4
n = 4
Loop iter: ev 5
Loop iter: ev 6
Loop iter: ev 7
Loop iter: ev 8
Loop iter: ev 9
Loop iter: ev 10
n = 10
Loop iter: ev 11
Loop iter: ev 12
Loop iter: ev 13
Loop iter: ev 14
Loop iter: ev 15
Loop iter: ev 16
Loop iter: ev 17
Loop iter: ev 18
Loop iter: ev 19
Loop iter: ev 20
Loop iter: ev 21
Loop iter: ev 22
n = 22
Loop iter: ev 23
Loop iter: ev 24
```

Reply:
What that is showing is that it is doubling the populated data everytime it resizes. This is also problematic as it means it will need double the computation necessary.

Reply:
Ah. Yes good point. I will try and rewrite it to do it efficiently.

Reply:
What do you think of the new change?